Achilles generation failing

Cannot Exectute "Generate" Achilles reports:

```
2021-01-25 08:24:07.708 DEBUG 110 --- [NodeScheduler-5] c.o.a.datanode.service.AchillesService   : Pulling image: hub.arachnenetwork.com/achilles:20190507
2021-01-25 08:24:07.793 ERROR 110 --- [tream-473388756] c.g.d.api.async.ResultCallbackTemplate   : Error during callback

javax.ws.rs.ProcessingException: org.glassfish.jersey.message.internal.HeaderUtils.asStringHeadersSingleValue(Ljavax/ws/rs/core/MultivaluedMap;Ljavax/ws/rs/core/Configuration;)Ljava/util/Map;
	at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:261) ~[jersey-client-2.25.1.jar!/:na]
	at org.glassfish.jersey.client.JerseyInvocation$1.call(JerseyInvocation.java:684) ~[jersey-client-2.25.1.jar!/:na]
	at org.glassfish.jersey.client.JerseyInvocation$1.call(JerseyInvocation.java:681) ~[jersey-client-2.25.1.jar!/:na]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315) ~[jersey-common-2.25.1.jar!/:na]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297) ~[jersey-common-2.25.1.jar!/:na]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:228) ~[jersey-common-2.25.1.jar!/:na]
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:444) ~[jersey-common-2.25.1.jar!/:na]
	at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:681) ~[jersey-client-2.25.1.jar!/:na]
	at com.github.dockerjava.jaxrs.JerseyDockerHttpClient.execute(JerseyDockerHttpClient.java:297) ~[docker-java-transport-jersey-3.2.6.jar!/:na]
	at com.github.dockerjava.core.DefaultInvocationBuilder.execute(DefaultInvocationBuilder.java:228) ~[docker-java-core-3.2.6.jar!/:na]
	at com.github.dockerjava.core.DefaultInvocationBuilder.lambda$executeAndStream$1(DefaultInvocationBuilder.java:269) ~[docker-java-core-3.2.6.jar!/:na]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_212]
Caused by: java.lang.NoSuchMethodError: org.glassfish.jersey.message.internal.HeaderUtils.asStringHeadersSingleValue(Ljavax/ws/rs/core/MultivaluedMap;Ljavax/ws/rs/core/Configuration;)Ljava/util/Map;
	at org.glassfish.jersey.apache.connector.ApacheConnector.writeOutBoundHeaders(ApacheConnector.java:653) ~[jersey-apache-connector-2.30.1.jar!/:na]
	at org.glassfish.jersey.apache.connector.ApacheConnector.apply(ApacheConnector.java:461) ~[jersey-apache-connector-2.30.1.jar!/:na]
	at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:252) ~[jersey-client-2.25.1.jar!/:na]
	... 11 common frames omitted

2021-01-25 08:24:07.817 ERROR 110 --- [NodeScheduler-5] c.o.a.datanode.service.AchillesService   : Achilles failed to execute

javax.ws.rs.ProcessingException: org.glassfish.jersey.message.internal.HeaderUtils.asStringHeadersSingleValue(Ljavax/ws/rs/core/MultivaluedMap;Ljavax/ws/rs/core/Configuration;)Ljava/util/Map;
	at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:261) ~[jersey-client-2.25.1.jar!/:na]
	at org.glassfish.jersey.client.JerseyInvocation$1.call(JerseyInvocation.java:684) ~[jersey-client-2.25.1.jar!/:na]
	at org.glassfish.jersey.client.JerseyInvocation$1.call(JerseyInvocation.java:681) ~[jersey-client-2.25.1.jar!/:na]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315) ~[jersey-common-2.25.1.jar!/:na]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297) ~[jersey-common-2.25.1.jar!/:na]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:228) ~[jersey-common-2.25.1.jar!/:na]
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:444) ~[jersey-common-2.25.1.jar!/:na]
	at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:681) ~[jersey-client-2.25.1.jar!/:na]
	at com.github.dockerjava.jaxrs.JerseyDockerHttpClient.execute(JerseyDockerHttpClient.java:297) ~[docker-java-transport-jersey-3.2.6.jar!/:na]
	at com.github.dockerjava.core.DefaultInvocationBuilder.execute(DefaultInvocationBuilder.java:228) ~[docker-java-core-3.2.6.jar!/:na]
	at com.github.dockerjava.core.DefaultInvocationBuilder.lambda$executeAndStream$1(DefaultInvocationBuilder.java:269) ~[docker-java-core-3.2.6.jar!/:na]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_212]
Caused by: java.lang.NoSuchMethodError: org.glassfish.jersey.message.internal.HeaderUtils.asStringHeadersSingleValue(Ljavax/ws/rs/core/MultivaluedMap;Ljavax/ws/rs/core/Configuration;)Ljava/util/Map;
	at org.glassfish.jersey.apache.connector.ApacheConnector.writeOutBoundHeaders(ApacheConnector.java:653) ~[jersey-apache-connector-2.30.1.jar!/:na]
	at org.glassfish.jersey.apache.connector.ApacheConnector.apply(ApacheConnector.java:461) ~[jersey-apache-connector-2.30.1.jar!/:na]
	at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:252) ~[jersey-client-2.25.1.jar!/:na]
	... 11 common frames omitted

--- a/pom.xml
+++ b/pom.xml
@@ -28,7 +28,7 @@
         <jjwt.version>0.10.5</jjwt.version>
         <postgresql.version>42.2.1</postgresql.version>
         <flywayVersion>4.2.0</flywayVersion>
-        <javaDockerVersion>3.2.6</javaDockerVersion>
+        <javaDockerVersion>3.2.7</javaDockerVersion>
         <springRetryVersion>1.2.4.RELEASE</springRetryVersion>
         <hikaricp.version>2.6.3</hikaricp.version>
         <build.number>${BUILD_NUMBER}</build.number>
@@ -159,11 +159,10 @@
             <artifactId>docker-java</artifactId>
             <version>${javaDockerVersion}</version>
         </dependency>
-
         <dependency>
-            <groupId>org.springframework.statemachine</groupId>
-            <artifactId>spring-statemachine-core</artifactId>
-            <version>1.1.1.RELEASE</version>
+            <groupId>com.github.docker-java</groupId>
+            <artifactId>docker-java-transport-httpclient5</artifactId>
+            <version>${javaDockerVersion}</version>
         </dependency>
 
         <!-- https://mvnrepository.com/artifact/commons-io/commons-io -->

--- a/src/main/java/com/odysseusinc/arachne/datanode/config/DockerConfig.java
+++ b/src/main/java/com/odysseusinc/arachne/datanode/config/DockerConfig.java
@@ -1,3 +1,5 @@
+package com.odysseusinc.arachne.datanode.config;
+
 /*
  *
  * Copyright 2018 Odysseus Data Services, inc.
@@ -20,11 +22,7 @@
  *
  */
 
-package com.odysseusinc.arachne.datanode.config;
-
-import com.github.dockerjava.api.DockerClient;
 import com.github.dockerjava.core.DefaultDockerClientConfig;
-import com.github.dockerjava.core.DockerClientBuilder;
 import com.github.dockerjava.core.DockerClientConfig;
 import com.odysseusinc.arachne.datanode.config.properties.DockerProperties;
 import org.springframework.boot.context.properties.EnableConfigurationProperties;
@@ -33,21 +31,17 @@ import org.springframework.context.annotation.Configuration;
 
 @Configuration
 @EnableConfigurationProperties(DockerProperties.class)
-public class DockerClientConfiguration {
-
+public class DockerConfig {
     @Bean
-    public DockerClient dockerClient(DockerProperties properties) {
+    public DockerClientConfig dockerClient(DockerProperties properties) {
 
-        DockerClientConfig config = DefaultDockerClientConfig.createDefaultConfigBuilder()
+        return DefaultDockerClientConfig.createDefaultConfigBuilder()
                 .withDockerHost(properties.getHost())
                 .withDockerTlsVerify(properties.isTlsVerify())
                 .withDockerCertPath(properties.getCertPath())
                 .withRegistryUrl(properties.getRegistry().getHost())
                 .withRegistryUsername(properties.getRegistry().getUsername())
                 .withRegistryPassword(properties.getRegistry().getPassword())
-                .build();
-
-        return DockerClientBuilder.getInstance(config)
                 .build();
     }
 }

--- a/src/main/java/com/odysseusinc/arachne/datanode/service/impl/AchillesServiceImpl.java
+++ b/src/main/java/com/odysseusinc/arachne/datanode/service/impl/AchillesServiceImpl.java
@@ -46,7 +46,6 @@ import com.github.dockerjava.api.command.WaitContainerResultCallback;
 import com.github.dockerjava.api.exception.NotFoundException;
 import com.github.dockerjava.api.model.AuthConfig;
 import com.github.dockerjava.api.model.Bind;
-import com.github.dockerjava.api.model.Frame;
 import com.github.dockerjava.api.model.HostConfig;
 import com.github.dockerjava.api.model.Volume;
 import com.github.dockerjava.core.DockerClientConfig;
@@ -609,12 +608,9 @@ public class AchillesServiceImpl implements AchillesService {
 
         ResultCallback.Adapter adapter = new ResultCallback.Adapter() {
             @Override
-            public void onNext(Object object) {
-                if (object instanceof Frame) {
-                    Frame item = (Frame) object;
-                    logEntries.add(item.toString());
-                    LOGGER.debug("{}", item);
-                }
+            public void onNext(Object item) {
+                logEntries.add(item.toString());
+                LOGGER.debug("{}", item);
             }
         };
 
@@ -624,12 +620,7 @@ public class AchillesServiceImpl implements AchillesService {
                 .withFollowStream(true)
                 .withTailAll()
                 .exec(adapter);
-        Integer statusCode = callback.awaitStatusCode();
-        try {
-            adapter.awaitCompletion(3, TimeUnit.SECONDS);
-        } catch (InterruptedException ignored) {
-        }
-        return statusCode;
+        return callback.awaitStatusCode();
     }
 
     private CreateContainerResponse buildContainer(DockerClient dockerClient, DataSource dataSource, Path workDir, String imageName) {

--- a/src/main/java/com/odysseusinc/arachne/datanode/service/impl/AchillesServiceImpl.java
+++ b/src/main/java/com/odysseusinc/arachne/datanode/service/impl/AchillesServiceImpl.java
@@ -38,16 +38,21 @@ import static com.odysseusinc.arachne.datanode.util.datasource.QueryProcessors.s
 import static java.lang.String.valueOf;
 
 import com.github.dockerjava.api.DockerClient;
+import com.github.dockerjava.api.async.ResultCallback;
 import com.github.dockerjava.api.command.CopyArchiveFromContainerCmd;
 import com.github.dockerjava.api.command.CreateContainerResponse;
+import com.github.dockerjava.api.command.PullImageResultCallback;
+import com.github.dockerjava.api.command.WaitContainerResultCallback;
 import com.github.dockerjava.api.exception.NotFoundException;
 import com.github.dockerjava.api.model.AuthConfig;
 import com.github.dockerjava.api.model.Bind;
 import com.github.dockerjava.api.model.Frame;
+import com.github.dockerjava.api.model.HostConfig;
 import com.github.dockerjava.api.model.Volume;
-import com.github.dockerjava.core.command.LogContainerResultCallback;
-import com.github.dockerjava.core.command.PullImageResultCallback;
-import com.github.dockerjava.core.command.WaitContainerResultCallback;
+import com.github.dockerjava.core.DockerClientConfig;
+import com.github.dockerjava.core.DockerClientImpl;
+import com.github.dockerjava.httpclient5.ApacheDockerHttpClient;
+import com.github.dockerjava.transport.DockerHttpClient;
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import com.odysseusinc.arachne.commons.api.v1.dto.CommonAchillesReportDTO;
 import com.odysseusinc.arachne.datanode.Constants;
@@ -55,7 +60,6 @@ import com.odysseusinc.arachne.datanode.config.properties.AchillesProperties;
 import com.odysseusinc.arachne.datanode.exception.AchillesJobInProgressException;
 import com.odysseusinc.arachne.datanode.exception.AchillesResultNotAvailableException;
 import com.odysseusinc.arachne.datanode.exception.ArachneSystemRuntimeException;
-import com.odysseusinc.arachne.datanode.exception.NotExistException;
 import com.odysseusinc.arachne.datanode.model.achilles.AchillesJob;
 import com.odysseusinc.arachne.datanode.model.achilles.AchillesJobSource;
 import com.odysseusinc.arachne.datanode.model.achilles.AchillesJobStatus;
@@ -87,6 +91,7 @@ import com.odysseusinc.arachne.datanode.util.datasource.ResultSetProcessor;
 import com.odysseusinc.arachne.datanode.util.datasource.ResultTransformers;
 import com.odysseusinc.arachne.datanode.util.datasource.ResultWriters;
 import com.odysseusinc.arachne.execution_engine_common.util.CommonFileUtils;
+
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.IOException;
@@ -119,6 +124,7 @@ import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 import javax.annotation.PostConstruct;
+
 import net.lingala.zip4j.exception.ZipException;
 import org.apache.commons.io.FileUtils;
 import org.apache.commons.io.IOUtils;
@@ -148,11 +154,11 @@ public class AchillesServiceImpl implements AchillesService {
     public static final String MEASUREMENT_TREEMAP_SQL = "classpath:/achilles/data/export/measurement/sqlMeasurementTreemap.sql";
     public static final String OBSERVATION_TREEMAP_SQL = "classpath:/achilles/data/export/observation/sqlObservationTreemap.sql";
     public static final String VISIT_SQL = "classpath:/achilles/data/export/visit/sqlVisitTreemap.sql";
-    protected static final Logger LOGGER = LoggerFactory.getLogger(AchillesService.class);
-    protected static final String DATA_NODE_NOT_EXISTS_EXCEPTION = "DataNode is not registered, please register";
     public static final String ACHILLES_RESULTS_EXCEPTION = "Achilles results is not available, table %s was not found";
+    protected static final Logger LOGGER = LoggerFactory.getLogger(AchillesServiceImpl.class);
+    protected static final String DATA_NODE_NOT_EXISTS_EXCEPTION = "DataNode is not registered, please register";
     private static final String ACHILLES_RESULTS_AVAILABLE_LOG = "Achilles results available at: {}";
-    protected final DockerClient dockerClient;
+    protected final DockerClientConfig dockerClientConfig;
     protected final AchillesProperties properties;
     protected final CentralUtil centralUtil;
     protected final RetryTemplate retryTemplate;
@@ -201,14 +207,14 @@ public class AchillesServiceImpl implements AchillesService {
 
     @SuppressWarnings("SpringJavaAutowiringInspection")
     @Autowired
-    public AchillesServiceImpl(DockerClient dockerClient,
+    public AchillesServiceImpl(DockerClientConfig dockerClientConfig,
                                AchillesProperties properties,
                                CentralUtil centralUtil,
                                @Qualifier("achillesRetryTemplate") RetryTemplate retryTemplate,
                                DataNodeService dataNodeService,
                                AchillesJobRepository achillesJobRepository) {
 
-        this.dockerClient = dockerClient;
+        this.dockerClientConfig = dockerClientConfig;
         this.properties = properties;
         this.centralUtil = centralUtil;
         this.retryTemplate = retryTemplate;
@@ -245,7 +251,7 @@ public class AchillesServiceImpl implements AchillesService {
             }
 
             updateJob(job, SUCCESSFUL);
-        } catch (Throwable e) {
+        } catch (Exception e) {
             LOGGER.error("Achilles failed to execute", e);
             updateJob(job, FAILED);
         } finally {
@@ -345,7 +351,7 @@ public class AchillesServiceImpl implements AchillesService {
             } finally {
                 executorService.shutdown();
             }
-        } catch (Throwable e) {
+        } catch (Exception e) {
             LOGGER.error("Failed to pull achilles results", e);
             updateJob(job, FAILED);
         }
@@ -503,12 +509,12 @@ public class AchillesServiceImpl implements AchillesService {
     }
 
     @Override
-    public List<CommonAchillesReportDTO> getAchillesReports() throws NotExistException {
+    public List<CommonAchillesReportDTO> getAchillesReports() {
 
         return centralSystemClient.listReports();
     }
 
-    private void sendResultToCentral(DataSource dataSource, Path results) throws IOException, NotExistException {
+    private void sendResultToCentral(DataSource dataSource, Path results) throws IOException {
 
         LOGGER.debug("Sending results to central: {}", centralHost);
         final File file = results.toFile();
@@ -541,21 +547,98 @@ public class AchillesServiceImpl implements AchillesService {
         );
     }
 
-    private Path runAchilles(final DataSource dataSource, AchillesJob job, Path workDir) throws IOException {
+    private DockerClient createDockerClient(DockerClientConfig dockerClientConfig) {
+        DockerHttpClient dockerHttpClient = new ApacheDockerHttpClient.Builder()
+                .dockerHost(dockerClientConfig.getDockerHost())
+                .sslConfig(dockerClientConfig.getSSLConfig())
+                .build();
+        return DockerClientImpl.getInstance(dockerClientConfig, dockerHttpClient);
+    }
+
+    private Path runAchilles(final DataSource dataSource, AchillesJob job, Path workDir) throws InterruptedException, IOException {
 
         LOGGER.debug("Achilles result directory: {}", workDir);
         String imageName = properties.getImageName();
-        LOGGER.debug("Pulling image: {}", imageName);
-        AuthConfig authConfig = new AuthConfig()
-                .withRegistryAddress(properties.getAuthConfig().getRegistryAddress())
-                .withUsername(properties.getAuthConfig().getUsername())
-                .withPassword(properties.getAuthConfig().getPassword());
-        dockerClient.pullImageCmd(imageName)
-                .withAuthConfig(authConfig)
-                .exec(new PullImageResultCallback())
-                .awaitSuccess();
+        List<String> logEntries = new LinkedList<>();
+
+        try (DockerClient dockerClient = createDockerClient(this.dockerClientConfig)) {
+            pullAchillesImage(dockerClient, imageName);
+            CreateContainerResponse container = buildContainer(dockerClient, dataSource, workDir, imageName);
+            String containerId = container.getId();
+            LOGGER.debug("Starting container: {}", containerId);
+            Integer statusCode = executeContainer(dockerClient, containerId, logEntries);
+            dockerClient.removeContainerCmd(containerId)
+                    .withRemoveVolumes(!LOGGER.isDebugEnabled())
+                    .exec();
+
+            LOGGER.debug("Achilles finished with status code: {}", statusCode);
+            if (statusCode == 0) {
+                Path jsonDir = createAchillesOutputResults(dataSource, workDir);
+                LOGGER.info(ACHILLES_RESULTS_AVAILABLE_LOG, jsonDir);
+                return jsonDir;
+            } else {
+                logAchillesError(dataSource, workDir, logEntries, dockerClient, containerId);
+                throw new IOException("Achilles exited with errors, lost results");
+            }
+
+        } finally {
+            job.setAchillesLog(logEntries.stream().map(s -> s.replaceAll("\u0000", "")).collect(Collectors.joining("\n")));
+            achillesJobRepository.save(job);
+        }
+    }
+
+    private void logAchillesError(DataSource dataSource, Path workDir, List<String> logEntries, DockerClient dockerClient, String containerId) throws IOException {
+        final CopyArchiveFromContainerCmd errorReport
+                = dockerClient.copyArchiveFromContainerCmd(containerId, "/opt/app/errorReport.txt");
+        try {
+            logEntries.addAll(IOUtils.readLines(errorReport.exec(), Charset.forName(StandardCharsets.UTF_8.name())));
+        } catch (NotFoundException e) {
+            logEntries.add("errorReport.txt does not exists");
+        }
+
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug(ACHILLES_RESULTS_AVAILABLE_LOG, createAchillesOutputResults(dataSource, workDir));
+        }
+    }
+
+    private Integer executeContainer(DockerClient dockerClient, String containerId, List<String> logEntries) {
+        dockerClient.startContainerCmd(containerId).exec();
+        LOGGER.debug("Container running: {}", containerId);
+        WaitContainerResultCallback callback = dockerClient.waitContainerCmd(containerId)
+                .exec(new WaitContainerResultCallback());
+
+        ResultCallback.Adapter adapter = new ResultCallback.Adapter() {
+            @Override
+            public void onNext(Object object) {
+                if (object instanceof Frame) {
+                    Frame item = (Frame) object;
+                    logEntries.add(item.toString());
+                    LOGGER.debug("{}", item);
+                }
+            }
+        };
+
+        dockerClient.logContainerCmd(containerId)
+                .withStdOut(true)
+                .withStdErr(true)
+                .withFollowStream(true)
+                .withTailAll()
+                .exec(adapter);
+        Integer statusCode = callback.awaitStatusCode();
+        try {
+            adapter.awaitCompletion(3, TimeUnit.SECONDS);
+        } catch (InterruptedException ignored) {
+        }
+        return statusCode;
+    }
+
+    private CreateContainerResponse buildContainer(DockerClient dockerClient, DataSource dataSource, Path workDir, String imageName) {
         LOGGER.debug("Creating container: {}", imageName);
         Volume outputVolume = new Volume("/opt/app/output");
+        HostConfig hostConfig = HostConfig.newHostConfig()
+                .withBinds(new Bind(getTempLocationOnHost(workDir), outputVolume))
+                .withNetworkMode(properties.getNetworkMode());
+
         CreateContainerResponse container = dockerClient.createContainerCmd(imageName)
                 .withEnv(
                         env(ACHILLES_SOURCE, dataSource.getName()),
@@ -567,64 +650,22 @@ public class AchillesServiceImpl implements AchillesService {
                         env(ACHILLES_NUM_THREADS, valueOf(1))
                 )
                 .withVolumes(outputVolume)
-                .withBinds(new Bind(getTempLocationOnHost(workDir), outputVolume))
-                .withNetworkMode(properties.getNetworkMode())
+                .withHostConfig(hostConfig)
                 .exec();
         LOGGER.debug("Container created: {}", container.getId());
-        final boolean debugEnabled = LOGGER.isDebugEnabled();
-        List<String> logEntries = new LinkedList<>();
-        try {
-            LOGGER.debug("Starting container: {}", container.getId());
-            dockerClient.startContainerCmd(container.getId()).exec();
-            LOGGER.debug("Container running: {}", container.getId());
-            WaitContainerResultCallback callback = dockerClient.waitContainerCmd(container.getId())
-                    .exec(new WaitContainerResultCallback());
-            LogContainerResultCallback logCallback = new LogContainerResultCallback() {
-                @Override
-                public void onNext(Frame item) {
+        return container;
+    }
 
-                    logEntries.add(item.toString());
-                    LOGGER.debug("{}", item);
-                }
-            };
-            dockerClient.logContainerCmd(container.getId())
-                    .withStdOut(true)
-                    .withStdErr(true)
-                    .withFollowStream(true)
-                    .withTailAll()
-                    .exec(logCallback);
-            Integer statusCode = callback.awaitStatusCode();
-            try {
-                logCallback.awaitCompletion(3, TimeUnit.SECONDS);
-            } catch (InterruptedException ignored) {
-            }
-            LOGGER.debug("Achilles finished with status code: {}", statusCode);
-            if (statusCode == 0) {
-                Path jsonDir = createAchillesOutputResults(dataSource, workDir);
-                LOGGER.info(ACHILLES_RESULTS_AVAILABLE_LOG, jsonDir);
-                return jsonDir;
-            } else {
-                String message = "Achilles exited with errors, lost results";
-                LOGGER.warn(message);
-                final CopyArchiveFromContainerCmd errorReport
-                        = dockerClient.copyArchiveFromContainerCmd(container.getId(), "/opt/app/errorReport.txt");
-                try {
-                    logEntries.addAll(IOUtils.readLines(errorReport.exec(), Charset.forName(StandardCharsets.UTF_8.name())));
-                } catch (NotFoundException e) {
-                    logEntries.add("errorReport.txt does not exists");
-                }
-                if (debugEnabled) {
-                    LOGGER.debug(ACHILLES_RESULTS_AVAILABLE_LOG, createAchillesOutputResults(dataSource, workDir));
-                }
-                throw new IOException(message);
-            }
-        } finally {
-            job.setAchillesLog(logEntries.stream().map(s -> s.replaceAll("\u0000", "")).collect(Collectors.joining("\n")));
-            achillesJobRepository.save(job);
-            dockerClient.removeContainerCmd(container.getId())
-                    .withRemoveVolumes(!debugEnabled)
-                    .exec();
-        }
+    private void pullAchillesImage(DockerClient dockerClient, String imageName) throws InterruptedException {
+        LOGGER.debug("Pulling image: {}", imageName);
+        AuthConfig authConfig = new AuthConfig()
+                .withRegistryAddress(properties.getAuthConfig().getRegistryAddress())
+                .withUsername(properties.getAuthConfig().getUsername())
+                .withPassword(properties.getAuthConfig().getPassword());
+        dockerClient.pullImageCmd(imageName)
+                .withAuthConfig(authConfig)
+                .exec(new PullImageResultCallback())
+                .awaitCompletion(5, TimeUnit.MINUTES);
     }
 
     private Path createAchillesOutputResults(DataSource dataSource, Path workDir) throws IOException {

